### PR TITLE
feature/deduple-middleware/#9

### PR DESCRIPTION
**Pull request by Uhsac**
#9

### CODE REVIEW

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# [Deduple middleware](https://github.com/botify-labs/botify-sdk-js-middlewares/blob/master/src/middlewares/dedupleMiddleware.js)
+
+When possible, this middleware will merge operation if they have the same params, this permit to reduce the number of call made to the API.
+
+If a request is merged it don't continue the middleware chain, so be carefull with the order.
+
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: Add sections
Middleware requirement
Operation options

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# [Deduple middleware](https://github.com/botify-labs/botify-sdk-js-middlewares/blob/master/src/middlewares/dedupleMiddleware.js)
+
+When possible, this middleware will merge operation if they have the same params, this permit to reduce the number of call made to the API.
+
+If a request is merged it don't continue the middleware chain, so be carefull with the order.
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: Middlewares chain of duplicate operations are interrupted. So you might consider middlewares order carefully.

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+)(baseSdk);
+
+const params = {
+  username: 'botify',
+  projectSlug: 'botify.com',
+  analyseSlug: 'foo',
+};
+
+// There is 2 same requests but with this middleware only one call will be made to the API
+sdk.AnalysesController.getQueryAggregate(
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: Use simple examples like
```JS
sdk.AnalysesController.getAnalysisSummary(params, (error, result) => {
   //Handle Result
}
```

File: src/middlewares/dedupleMiddleware.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * @return {Middleware}
+ */
+export default function dedupleMiddleware() {
+  const currentOperations = {};
+
+  return next => function(params, callback, options) {
+    const hash = objectHash({
+      params,
+      options,
```
Comment (zallek) [inline src/middlewares/dedupleMiddleware.js]: add controllerId and operationId to hash (same for batchmiddleware)

File: src/middlewares/dedupleMiddleware.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    if (currentOperations[hash]) {
+      currentOperations[hash].push(callback);
+    } else {
+      currentOperations[hash] = [callback];
+      next(
+        params,
+        (error, result) => {
+          currentOperations[hash].forEach(cb => cb(error, result));
+          delete currentOperations[hash];
```
Comment (zallek) [inline src/middlewares/dedupleMiddleware.js]: currentOperations[hash] = null; for compatibility

File: test/middlewares/dedupleMiddleware.spec.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    middleware(spiedOperation)(1, spiedCallback);
+    middleware(spiedOperation)(1, spiedCallback);
+    setImmediate(() => {
+      chai.expect(spiedOperation.callCount).to.be.equal(1);
+      chai.expect(spiedCallback.callCount).to.be.equal(2);
+      done();
+    });
+  });
+
+  it('must call operation two time if multiple call with different parameters', done => {
```
Comment (zallek) [inline test/middlewares/dedupleMiddleware.spec.js]: add tests with controller and operation ids

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    //Handle Result
+  }
+);
+sdk.AnalysesController.getQueryAggregate(
+  {...params, queries: Array<Query>},
+  (error, result) => {
+    //Handle Result
+  }
+);
+```
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: Update recommanded setup in introduction

**Comment by Uhsac** [comment]
@zallek I update batchMiddleware hash in this PR, do you want me to do this in a new one ?

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+// There is 2 same requests but with this middleware only one call will be made to the API
+sdk.AnalysesController.getAnalysisSummary(params, (error, result) => {
+   //Handle Result
+});
+sdk.AnalysesController.getAnalysisSummary(params, (error, result) => {
+   //Handle Result
+});
+```
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: Update recommanded setup in introduction

File: docs/middlewares/dedupleMiddleware.docs.md [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+## Middleware requirement
+none
+
+## Operation options
+none
+
+## Usage
+``` javascript
+import { applyMiddleware, middlewares } from 'botify-sdk-middlewares';
+const { dedupleMiddleware } = middlewares;
```
Comment (zallek) [inline docs/middlewares/dedupleMiddleware.docs.md]: there is no more `middlewares` key

File: src/middlewares/dedupleMiddleware.js [hunk 10]
```diff
@@ -0,0 +1,31 @@
+import objectHash from 'object-hash';
+
+/**
+ * @return {Middleware}
```
Comment (zallek) [inline src/middlewares/dedupleMiddleware.js]: hum it doesn't really returns a middlewares, this this the middlewares but when you have middlewares options you need to create a factory that returns a middleware

File: src/middlewares/batchMiddleware.js [hunk 11]
```diff
@@ -133,14 +134,17 @@ export default function({
...
       if (!batchOperation) {
-        return next(...arguments);
+        return next(params, callback, options);
       }
 
       const hash = objectHash({
         commonParams: pick(params, batchOperation.commonKeys),
+        options,
+        controllerId,
         operationId,
```
Comment (zallek) [inline src/middlewares/batchMiddleware.js]: options should be sent by queue.

**Pull request merged**